[Feature:Submission] Late Day database triggers

jportorreal00 opened the pull request:
### Please check if the PR fulfills these requirements:

* [ ] Tests for the changes have been added/updated (if possible)
* [ ] Documentation has been updated/added if relevant

### What is the current behavior?
Late days are currently calculated within the LateDays object via php. Current late days use previous late day events as context in order to determine the #of late days remaining and # of late days used.

### What is the new behavior?
Late day information will now be stored in a table. (Table is similar to Late day usage table). This will make accessing late day information a lot faster since we will not have to calculate it from scratch every time.

Late days will now be automatically reset and recalculated using triggers within the databases. Cases where this will happen:

- submit new gradeable
- gradeable version change
- drop submission
- late day (allowed table) addition
- late day (allowed table) deletion
- gradeable late days allowed change
- gradeable due date change
- gradeable due date disabling/enabling
- gradeable late days disabling/enabling
- gradeable deletion
- add new late day extension
- edit existing late day extension
- delete late day extension
- change default late days (this is the only **manual** table reset: inside the php code)


### Other information?
<!-- Is this a breaking change? -->
<!-- How did you test -->
I tested locally on two courses. I ran over each of the scenarios listed and they work as expected (using pgAdmin)

The changes to twig files were due to the changes in the Late Day Usage data. Ex:
![image](https://user-images.githubusercontent.com/51247866/152180489-119c8e51-4d86-4367-816d-cfd8aca167fa.png)

williamjallen commented on migration/migrator/migrations/course/20211117160515_add_late_day_triggers.py:
```diff
@@ -0,0 +1,376 @@
+        AS $$
+    #variable_conflict use_variable
+    DECLARE
+    latestDate timestamp without time zone ;
```
Is there a particular reason why this doesn't have a timezone?

jportorreal00 replied:
good point

williamjallen commented on site/app/controllers/admin/ConfigurationController.php:
```diff
@@ -206,6 +206,11 @@ public function updateConfiguration(): MultiResponse {
         }
 
+        // All late day cache now invalid
+        if ($name == 'default_student_late_days') {
```
It would be better to use `===` here instead of `==`

williamjallen commented on site/app/libraries/database/DatabaseQueries.php:
```diff
@@ -1407,6 +1407,102 @@ public function getLateDayInformation($user_id) {
+                $late_day_events[$row['g_id']] = $row;
+            }
+            else { // Late day update event
+                $late_day_events[$index++] = $row;
```
Is the intention here to use pre- or post-increment for `$index` here?

williamjallen replied:
Hmm, if I'm understanding what this code does properly, it seems *very* prone to someone making changes in the future and accidentally breaking things.  I would recommend doing something like the following just for clarity.
```
$late_day_events[$index] = $row;
$index++; 
```

williamjallen commented on site/app/libraries/database/DatabaseQueries.php:
```diff
@@ -1407,6 +1407,102 @@ public function getLateDayInformation($user_id) {
+        return $return;
+    }
+
+    public function getLateDayCacheForUser($user_id) {
```
It would be good to add type hints to the parameters and return value for this function

williamjallen replied:
Perhaps Matt can correct me if I'm wrong here but in my opinion, it would be better to use [this](https://www.php.net/manual/en/language.types.declarations.php) type of type hinting.  Using the above method instead of comments, any parameters that are not of the proper type will result in PHP errors in the development environment which helps to catch bugs earlier in my experience.

jportorreal00 replied:
I see what you mean. In that case do you think I should remove the docstring or keep it?

williamjallen replied:
For simple things where the docstring duplicates the information provided by the type hint, I would suggest removing it because it simply clutters the code.  For things where you have specified the return type in more detail, I would argue that it would be good to keep it so that someone can quickly reference it in the future should they wander into this part of the code and wonder what that function returns.  

Matt may have a better suggestion here though.

jportorreal00 replied:
I see some other functions/constructors with the duplicate type hinting, so I will keep it for now. Thanks!

williamjallen commented on site/app/libraries/database/DatabaseQueries.php:
```diff
@@ -1407,6 +1407,102 @@ public function getLateDayInformation($user_id) {
         return $this->course_db->rows();
     }
 
+    public function getLateDayCache() {
```
It would be good to add type hints for the return value here

williamjallen commented on site/app/libraries/database/DatabaseQueries.php:
```diff
@@ -1407,6 +1407,102 @@ public function getLateDayInformation($user_id) {
+        return $late_day_events;
+    }
+
+    public function getLateDayCacheForUserGradeable($user_id, $g_id) {
```
It would be good to add type hints to the parameters and return value for this function

williamjallen commented on site/app/libraries/database/DatabaseQueries.php:
```diff
@@ -1407,6 +1407,102 @@ public function getLateDayInformation($user_id) {
+        return $row;
+    }
+
+    public function generateLateDayCacheForUsers() {
```
It would be good to add type hints to the return value for this function

williamjallen commented on site/app/libraries/database/DatabaseQueries.php:
```diff
@@ -1407,6 +1407,102 @@ public function getLateDayInformation($user_id) {
+        $this->course_db->query($query, $params);
+    }
+
+    public function generateLateDayCacheForUser($user_id) {
```
Same comment here about type-hinting

williamjallen commented on site/app/models/gradeable/LateDayInfo.php:
```diff
@@ -71,6 +100,51 @@ public function toArray() {
+     * Gets information about this late day event in the format for the cache table
+     * @return array
+     */
+    public function generateEventInfo() {
```
Same comment here about type-hinting (applies to all of the following functions)

williamjallen commented on site/app/models/gradeable/LateDayInfo.php:
```diff
@@ -161,43 +236,64 @@ public function getStatusMessage() {
      */
     public function getGradeInquiryCount() {
-        return $this->graded_gradeable->getGradeInquiryCount();
+        return $this->graded_gradeable !== null ? $this->graded_gradeable->getGradeInquiryCount() : null;
```
It would be better to use the the null coalescing (`??`) operator for this I believe.

jportorreal00 replied:
Since I'm checking $graded_gradeable I cant use the coalescing operator. It might be null, I wont be sure if I can use `getGradeInquiryCount()`. I can potentially shorten this with `$this->graded_gradeable?->getGradeInquiryCount() ?? null;` but we would need PHP v8 and up tp use the `?->` operator.

williamjallen commented on site/tests/app/models/gradeable/LateDaysTester.php:
```diff
@@ -94,20 +95,81 @@ private function mockGradedGradeables() {
+    }
 
-        return [$on_time, $on_time_exceptions, $late, $late_exception, $bad_for_gradeable, $bad_for_term, $on_time1, $late1];
+    private function createMockCache($graded_gradeables, $late_days, $updates) {
```
It would be good to add type hinting here

williamjallen reviewed (changes requested):
Code looks good in general, but I think that it would be good to add type hinting in a bunch of places to ensure (at least in the development environment) that errors are thrown if unexpected parameters/return types are provided.
Example with type hinting on two parameters and the return type:
```
public function foo(string param1, int param2): array {
    ...
    return [1, 2, 3];
}
```

williamjallen commented on site/app/models/gradeable/LateDayInfo.php:
```diff
@@ -161,43 +236,64 @@ public function getStatusMessage() {
-            return 0;
-        }
-        // clamp the days charged to be the days late minus exceptions above zero.
-        return max(0, min($this->getDaysLate(), $this->getLateDaysAllowed()) - $this->getLateDayException());
+        return -$this->late_days_change;
+    }
+
+    /**
+     * Gets the number of late days increased from this event
+     * @return int
+     */
+    public function getLateDaysChange() {
+        return $this->late_days_change;
     }
```
If I was reading through the code and encountered `getLateDaysCharged()` I would not immediately expect this function to return a negative number.  Rather, I would expect it to return a positive number representing the number of late days charged.  Similarly, it seems rather confusing to have two functions which return the exact same thing.  Is there a reason for this or should `getLateDaysCharged()` be eliminated altogether?

Just to be clear, this isn't a requested change--just food for thought in case changes are desired.

jportorreal00 replied:
Very good point. I opted to max out the value at 0. I think it could provide better functionality for future changes. I think only late gradeables should return a significant value.

williamjallen commented on site/app/templates/grading/electronic/StudentInformationPanel.twig:
```diff
@@ -40,12 +40,60 @@
+    $(document).ready(function() {
+        // Disable any nav-bar buttons with no content
+        $('.nav-bar').each(function () {
+            if ($('#page_' + this.id.split('_')[1] + '_content').length == 0) {
```
It would be better to use `===` here

jportorreal00 replied:
oops this wasnt meant to be in this branch

williamjallen commented:
I like the most recent changes, but I wonder if this PR is the best PR to include them in.  Would it be better to hold off on making changes to the TA grading interface given that this PR is mostly setting up the triggers?  The new changes could be made as a new PR, or added to one of the other open PRs--I'm not sure which would be best in this case.

jportorreal00 commented:
> I like the most recent changes, but I wonder if this PR is the best PR to include them in. Would it be better to hold off on making changes to the TA grading interface given that this PR is mostly setting up the triggers? The new changes could be made as a new PR, or added to one of the other open PRs--I'm not sure which would be best in this case.

Yes definitely!! It was meant to be part of #7702, but i goofed up the commits somehow. Sorry about that.

MasterOdin commented:
@jportorreal00 Please try to avoid doing force pushes on reviewed PRs that completely change the history. It makes it hard to know on what commits the reviews were made, and if you made a ton of changes after a review.

jportorreal00 commented:
> @jportorreal00 Please try to avoid doing force pushes on reviewed PRs that completely change the history. It makes it hard to know on what commits the reviews were made, and if you made a ton of changes after a review.

Oh shoot sorry about that! I was pulling from master and didn’t want to deal with merge conflicts. I’ll avoid doing that from now on.

MasterOdin commented:
Meant to add, do you know what (if any) commits you added after @williamjallen's approval?

jportorreal00 commented:
He approved on 3/2. I added a commit on 3/3 that belonged to #7702. I realized it was wrong so I reverted to the previous commit. So the current state is the commits that William had the chance to look over

jportorreal00 commented:
@MasterOdin heres my local git log

![image](https://user-images.githubusercontent.com/51247866/158295699-3b151108-4075-4416-9be8-310afe1018db.png)

jportorreal00 commented:
@williamjallen @MasterOdin There was a new commit added to fix a bug with the Reports for CSVs and rainbow grades. Only electronic gradeables with late days allowed (and not in the future) support late day cache. Any other gradeables don't effect late days, there for they don't have any late day info.
